oap-server/server-query-plugin/zipkin-query-plugin/src/main/java/org/apache/skywalking/oap/query/zipkin/handler/ZipkinQueryHandler.java

```diff
@@ -304,6 +306,16 @@ private byte[] encodeTraces(List<List<Span>> traces) {
         return buff.array();
     }
 
+    private void appendEventsToTraces(List<List<Span>> traces) throws IOException {
+        for (List<Span> spans : traces) {
+            if (CollectionUtils.isEmpty(spans)) {
+                continue;
+            }
+
+            appendEvents(spans, getSpanAttachedEventQueryDAO().querySpanAttachedEvents(SpanAttachedEventTraceType.ZIPKIN, spans.get(0).traceId()));
```

Comment:
Should consider a list query rather than this.
10-20 iterations would be very slow.

Reply:
Updated, please re-check.